### ISSUE TITLE
tinyZenpage plugin: fails to add photos to pages with (') in titles

### ISSUE THREAD

**Issue by kuz-z-zma** [zenphoto/zenphoto#1472]
The issue: 
when adding images to news or pages via tinyZenpage plugin, image won't be added if the image Title contains an apostrophe (').
- there is no error, adding just silently fails
- all images are present in the selector window
- other parameters seem to have no effect - any image size, any included links won't affect the outcome.

The issue is not new, based on my notes - it was present as far back as January 2024.

**Comment by acrylian** [comment]
> The issue is not new, based on my notes - it was present as far back as January 2024.

And why didn't you tell earlier? We are about to release 1.6.5 actually by now… Anyways, I suspect this will be a JS issue.

**Comment by kuz-z-zma** [comment]
Why not earlier: couldn't test and isolate the issue, not enough data and no time to investigate then.
Who knows what's at fault?
I first thought the issue was with folders or filenames, or my tinyMCE configuration, or even some browser blockage, but apparently not.

**Comment by acrylian** [comment]
Okay, "no time" is certainly a very valid reason I understand ;-) We of course could have helped testing/finding it. 

I just suspected JS as that is often very picky and this might be an single quote/quote issue since the char you used above is a actually single quote code wise and not a "apostrophe" typographically (typographically it's also not a quote but that's very specific ;-))

**Comment by kuz-z-zma** [comment]
I just prefer to report issues with some testing already done, at least to make sure that's an actual issue with Zenphoto and not somewhere on my end. Seems to be wasteful to have an ability and not to use it to make life easier for others.

I've encountered this issue sometime earlier this year, seemed weird and I made a note to look into it, but only after I've started to regularly update one of my sites with news and work on it's pages, was I able to find other affected photos and make a connection.

Probably should test other special characters too, might have the same effect.

**Comment by acrylian** [comment]
Sure, I understand. It's good to be sure it wasn't a glitch before reporting. That "apostrophe" is technically not a special char and I am pretty sure it is some esaping quote issue. But some test would of course be good. I probably won't be able to really work on something much before Monday.

**Comment by kuz-z-zma** [comment]
I've tested special characters in Titles:

-  ' and " cause failure to add image, 
- !@#$%^&*?/| and other - don't cause problems.
- < and > get converted to `&lt;` and `&rt;` and don't cause problems

**Comment by acrylian** [comment]
There is a possible fix in the repo now.

**Comment by kuz-z-zma** [comment]
@acrylian I've tested the fix:
- images with (') in titles are now added
- but images with (") in titles still not added via tinyZenpage dialog

**Comment by acrylian** [comment]
Totally missed that inch ("quote") is also an issue being added in the later comment. Will try to take another look tomorrow…

**Comment by bic-ed** [comment]
As told to @acrylian and @fretzl by mail, I have a fix which is working on this issue, avoiding this silent fail on inserting items, but there is still a minor issue I have no time to further investigate at the moment. The title attribute for the link inserted by the plugin gets truncated at the first occurrence of the `'` character. I'll push the fix as is anyway for now.

**Comment by bic-ed** [comment]
For the title attribute issue, a solution seems to be escaping again in JavaScript the plainimgtitle and plainalbumtitle variables, which are used for the title attribute. We could use a function like the following, instead of the current one which only escapes < and > :
```javascript
function escapeHtml(text) {
  var map = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#039;'
  };
  
  return text.replace(/[&<>"']/g, function(m) { return map[m]; });
}
``` 

Or a better one if any.

**Comment by bic-ed** [comment]
Or a simplified version, since the only character that is creating problems is the single quote `'`, because it is used inside single quotes. Such as
```javascript
plainimgtitle = plainimgtitle.replace(/'/g, "&#039;");
```

**Comment by kuz-z-zma** [comment]
@bic-ed did a test with your fix.
Photos with (') and (") seem to be added now without problems via tinyZenpage: I've played with various sizes and links.

Finally, I've put this random string in title: **Someone will remember us"'|?>\/<>!@#$%^&*()**
The image is added, but special characters result in this:
![image](https://github.com/user-attachments/assets/68f9aed9-7223-4f08-811c-92e36671ca50)

Code-wise it's transformed into:
`<p><a class="zenpage_imagelink" title="Someone will remember us&quot;'|?&gt;\/&lt;&gt;!@#$%^&amp;*()" href="/index.php?album=other-formats&amp;image=tumblr_sapho.jpg"><img src="/cache/other-formats/tumblr_sapho_240_thumb.jpg?cached=1713443087" alt="Someone will remember us&quot;" />\/&lt;&gt;!@#$%^&amp;*()' class='zenpage_thumb' loading='lazy' /&gt;</a></p>`

**Comment by bic-ed** [comment]
Hey, you killed a fly with a gun, but this is actually a good find, thanks! I'm on my phone now, so I can't be 100% sure and cannot test anything, but it seems to me that it is not the title attribute to be broken but the alt one. Eventually we missed to check that as well. Tomorrow morning I will probably fix it.

**Comment by bic-ed** [comment]
Yes, it was the alt attribute.
@kuz-z-zma, can you test again please?

**Comment by kuz-z-zma** [comment]
@bic-ed: I've tested the new fix and everything seems to work as expected with whatever I throw at it.

Thank you!

**Comment by kuz-z-zma** [comment]
@bic-ed @acrylian 

I've added my "random string of death with special characters" (**"'|?>/<>!@#$%^&*()**) to the description of the photo - and this also causes silent failure to add image via tinyZenpage.
Sorry, didn't think to test this too earlier...

**Comment by bic-ed** [comment]
We can make it work as well. The problematic char is again the single quote `'`, and for descriptions we don't have yet implemented the addlashes encoding on php side, along with the subsequent replacing `albumdesc.replace(/'/g, "&#039;")` on javascript side.
What remains unclear to me is why our html_encode() function seems to skip encoding the single quote. I've looked at the flags passed to the native function htmlspecialchars() and the right flag ENT_QUOTES seems to be there.

**Comment by kuz-z-zma** [comment]
@bic-ed: tested and all works now with no issues!
